File: java/src/org/smc/inputmethod/indic/settings/AdvancedSettingsFragment.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.media.AudioManager;
import android.os.Bundle;

import androidx.preference.Preference;

import com.android.inputmethod.latin.AudioAndHapticFeedbackManager;
import com.android.inputmethod.latin.R;
import com.android.inputmethod.latin.SystemBroadcastReceiver;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

/**
 * "Advanced" settings sub screen.
 *
 * This settings sub screen handles the following advanced preferences.
 * - Key popup dismiss delay
 * - Keypress vibration duration
 * - Keypress sound volume
 * - Show app icon
 * - Improve keyboard
 * - Debug settings
 */
public final class AdvancedSettingsFragment extends SubScreenFragment {
    @Override
    public void onCreatePreferences(final Bundle savedInstanceState, final String rootKey) {
        super.onCreatePreferences(savedInstanceState, rootKey);
        addPreferencesFromResource(R.xml.prefs_screen_advanced);

        final Resources res = getResources();
        final Context context = getActivity();

        // When we are called from the Settings application but we are not already running, some
        // singleton and utility classes may not have been initialized.  We have to call
        // initialization method of these classes here. See {@link LatinIME#onCreate()}.
        AudioAndHapticFeedbackManager.init(context);

        final SharedPreferences prefs = getPreferenceManager().getSharedPreferences();

        if (!Settings.isInternal(prefs)) {
            removePreference(Settings.SCREEN_DEBUG);
        }

        if (!AudioAndHapticFeedbackManager.getInstance().hasVibrator()) {
            removePreference(Settings.PREF_VIBRATION_DURATION_SETTINGS);
        }

        setupKeypressVibrationDurationSettings();
        setupKeypressSoundVolumeSettings();
        setupKeyLongpressTimeoutSettings();
        setupVarnamPreference();
        refreshEnablingsOfKeypressSoundAndVibrationSettings();
    }

    private static final String VARNAM_PACKAGE = "org.smc.inputmethod.indic.varnam";
    private static final String VARNAM_ACTIVITY = VARNAM_PACKAGE + ".MainActivity";

    private void setupVarnamPreference() {
        final Preference varnamPref = findPreference("screen_varnam");
        if (varnamPref == null) {
            return;
        }
        varnamPref.setOnPreferenceClickListener(preference -> {
            final Context context = getActivity();
            if (context == null) {
                return true;
            }
            if (isPackageInstalled(context, VARNAM_PACKAGE)) {
                final Intent intent = new Intent();
                intent.setClassName(VARNAM_PACKAGE, VARNAM_ACTIVITY);
                try {
                    startActivity(intent);
                } catch (final ActivityNotFoundException e) {
                    showVarnamInstallDialog(context);
                }
            } else {
                showVarnamInstallDialog(context);
            }
            return true;
        });
    }

    private static boolean isPackageInstalled(final Context context, final String packageName) {
        try {
            context.getPackageManager().getPackageInfo(packageName, 0);
            return true;
        } catch (final PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    private void showVarnamInstallDialog(final Context context) {
        new MaterialAlertDialogBuilder(context)
                .setTitle(R.string.varnam_not_installed_title)
                .setMessage(R.string.varnam_not_installed_message)
                .setPositiveButton(R.string.varnam_not_installed_dismiss, null)
                .show();
    }

    @Override
    public void onSharedPreferenceChanged(final SharedPreferences prefs, final String key) {
        if (key.equals(Settings.PREF_SHOW_SETUP_WIZARD_ICON)) {
            SystemBroadcastReceiver.toggleAppIcon(getActivity());
        }
        refreshEnablingsOfKeypressSoundAndVibrationSettings();
    }

    private void refreshEnablingsOfKeypressSoundAndVibrationSettings() {
        final SharedPreferences prefs = getSharedPreferences();
        final Resources res = getResources();
        setPreferenceEnabled(Settings.PREF_VIBRATION_DURATION_SETTINGS,
                Settings.readVibrationEnabled(prefs, res));
        setPreferenceEnabled(Settings.PREF_KEYPRESS_SOUND_VOLUME,
                Settings.readKeypressSoundEnabled(prefs, res));
    }

    private void setupKeypressVibrationDurationSettings() {
        final SeekBarDialogPreference pref = (SeekBarDialogPreference)findPreference(
                Settings.PREF_VIBRATION_DURATION_SETTINGS);
        if (pref == null) {
            return;
        }
        final SharedPreferences prefs = getSharedPreferences();
        final Resources res = getResources();
        pref.setInterface(new SeekBarDialogPreference.ValueProxy() {
            @Override
            public void writeValue(final int value, final String key) {
                prefs.edit().putInt(key, value).apply();
            }

            @Override
            public void writeDefaultValue(final String key) {
                prefs.edit().remove(key).apply();
            }

            @Override
            public int readValue(final String key) {
                return Settings.readKeypressVibrationDuration(prefs, res);
            }

            @Override
            public int readDefaultValue(final String key) {
                return Settings.readDefaultKeypressVibrationDuration(res);
            }

            @Override
            public void feedbackValue(final int value) {
                AudioAndHapticFeedbackManager.getInstance().vibrate(value);
            }

            @Override
            public String getValueText(final int value) {
                if (value < 0) {
                    return res.getString(R.string.settings_system_default);
                }
                return res.getString(R.string.abbreviation_unit_milliseconds, value);
            }
        });
    }

    private void setupKeypressSoundVolumeSettings() {
        final SeekBarDialogPreference pref = (SeekBarDialogPreference)findPreference(
                Settings.PREF_KEYPRESS_SOUND_VOLUME);
        if (pref == null) {
            return;
        }
        final SharedPreferences prefs = getSharedPreferences();
        final Resources res = getResources();
        final AudioManager am = (AudioManager)getActivity().getSystemService(Context.AUDIO_SERVICE);
        pref.setInterface(new SeekBarDialogPreference.ValueProxy() {
            private static final float PERCENTAGE_FLOAT = 100.0f;

            private float getValueFromPercentage(final int percentage) {
                return percentage / PERCENTAGE_FLOAT;
            }

            private int getPercentageFromValue(final float floatValue) {
                return (int)(floatValue * PERCENTAGE_FLOAT);
            }

            @Override
            public void writeValue(final int value, final String key) {
                prefs.edit().putFloat(key, getValueFromPercentage(value)).apply();
            }

            @Override
            public void writeDefaultValue(final String key) {
                prefs.edit().remove(key).apply();
            }

            @Override
            public int readValue(final String key) {
                return getPercentageFromValue(Settings.readKeypressSoundVolume(prefs, res));
            }

            @Override
            public int readDefaultValue(final String key) {
                return getPercentageFromValue(Settings.readDefaultKeypressSoundVolume(res));
            }

            @Override
            public String getValueText(final int value) {
                if (value < 0) {
                    return res.getString(R.string.settings_system_default);
                }
                return Integer.toString(value);
            }

            @Override
            public void feedbackValue(final int value) {
                am.playSoundEffect(
                        AudioManager.FX_KEYPRESS_STANDARD, getValueFromPercentage(value));
            }
        });
    }

    private void setupKeyLongpressTimeoutSettings() {
        final SharedPreferences prefs = getSharedPreferences();
        final Resources res = getResources();
        final SeekBarDialogPreference pref = (SeekBarDialogPreference)findPreference(
                Settings.PREF_KEY_LONGPRESS_TIMEOUT);
        if (pref == null) {
            return;
        }
        pref.setInterface(new SeekBarDialogPreference.ValueProxy() {
            @Override
            public void writeValue(final int value, final String key) {
                prefs.edit().putInt(key, value).apply();
            }

            @Override
            public void writeDefaultValue(final String key) {
                prefs.edit().remove(key).apply();
            }

            @Override
            public int readValue(final String key) {
                return Settings.readKeyLongpressTimeout(prefs, res);
            }

            @Override
            public int readDefaultValue(final String key) {
                return Settings.readDefaultKeyLongpressTimeout(res);
            }

            @Override
            public String getValueText(final int value) {
                return res.getString(R.string.abbreviation_unit_milliseconds, value);
            }

            @Override
            public void feedbackValue(final int value) {}
        });
    }
}

File: java/src/org/smc/inputmethod/indic/settings/AdvancedSettingsFragment.kt
/*
 * Copyright 2026, Jishnu Mohan <jishnu@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings

import android.content.Intent
import android.content.SharedPreferences
import android.media.AudioManager
import android.os.Bundle

import androidx.core.content.getSystemService
import androidx.preference.Preference

import com.android.inputmethod.latin.AudioAndHapticFeedbackManager
import com.android.inputmethod.latin.R
import com.android.inputmethod.latin.SystemBroadcastReceiver

import org.smc.inputmethod.indic.setup.SetupWizardActivity

/**
 * "Advanced" settings sub screen: key-popup dismiss delay, keypress vibration duration, keypress
 * sound volume, show-app-icon, improve-keyboard and the debug settings entry.
 */
class AdvancedSettingsFragment : SubScreenFragment() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        super.onCreatePreferences(savedInstanceState, rootKey)
        addPreferencesFromResource(R.xml.prefs_screen_advanced)

        // When launched from the Settings app while the IME is not running, some singletons may
        // not have been initialized yet. See LatinIME.onCreate().
        AudioAndHapticFeedbackManager.init(requireContext())

        if (!Settings.isInternal(sharedPreferences)) {
            removePreference(Settings.SCREEN_DEBUG)
        }
        if (!AudioAndHapticFeedbackManager.getInstance().hasVibrator()) {
            removePreference(Settings.PREF_VIBRATION_DURATION_SETTINGS)
        }

        setupKeypressVibrationDurationSettings()
        setupKeypressSoundVolumeSettings()
        setupKeyLongpressTimeoutSettings()
        refreshEnablingsOfKeypressSoundAndVibrationSettings()

        findPreference<Preference>(Settings.PREF_RERUN_SETUP)?.setOnPreferenceClickListener {
            startActivity(
                Intent(requireContext(), SetupWizardActivity::class.java)
                    .putExtra(SetupWizardActivity.EXTRA_FORCE_SETUP, true)
            )
            true
        }
    }

    override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String?) {
        if (key == Settings.PREF_SHOW_SETUP_WIZARD_ICON) {
            SystemBroadcastReceiver.toggleAppIcon(activity)
        }
        refreshEnablingsOfKeypressSoundAndVibrationSettings()
    }

    private fun setupKeypressVibrationDurationSettings() {
        val pref = findPreference<SeekBarDialogPreference>(
            Settings.PREF_VIBRATION_DURATION_SETTINGS
        ) ?: return
        val prefs = sharedPreferences
        val res = resources
        pref.bindValueProxy(
            prefs,
            read = { Settings.readKeypressVibrationDuration(prefs, res) },
            readDefault = { Settings.readDefaultKeypressVibrationDuration(res) },
            text = {
                if (it < 0) res.getString(R.string.settings_system_default)
                else res.getString(R.string.abbreviation_unit_milliseconds, it)
            },
            feedback = { AudioAndHapticFeedbackManager.getInstance().vibrate(it.toLong()) }
        )
    }

    private fun setupKeypressSoundVolumeSettings() {
        val pref = findPreference<SeekBarDialogPreference>(
            Settings.PREF_KEYPRESS_SOUND_VOLUME
        ) ?: return
        val prefs = sharedPreferences
        val res = resources
        val am = requireActivity().getSystemService<AudioManager>()!!
        pref.bindValueProxy(
            prefs,
            storeAsFraction = true,
            read = { (Settings.readKeypressSoundVolume(prefs, res) * 100f).toInt() },
            readDefault = { (Settings.readDefaultKeypressSoundVolume(res) * 100f).toInt() },
            text = { if (it < 0) res.getString(R.string.settings_system_default) else it.toString() },
            feedback = { am.playSoundEffect(AudioManager.FX_KEYPRESS_STANDARD, it / 100f) }
        )
    }

    private fun setupKeyLongpressTimeoutSettings() {
        val prefs = sharedPreferences
        val res = resources
        val pref = findPreference<SeekBarDialogPreference>(
            Settings.PREF_KEY_LONGPRESS_TIMEOUT
        ) ?: return
        pref.bindValueProxy(
            prefs,
            read = { Settings.readKeyLongpressTimeout(prefs, res) },
            readDefault = { Settings.readDefaultKeyLongpressTimeout(res) },
            text = { res.getString(R.string.abbreviation_unit_milliseconds, it) }
        )
    }
}
